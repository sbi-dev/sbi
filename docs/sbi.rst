.. currentmodule:: sbi

API Reference
===========================

.. toctree::
   :maxdepth: 1

   reference/sbi.inference
   reference/sbi.models
   reference/sbi.posteriors
   reference/sbi.potentials
   reference/sbi.analysis
