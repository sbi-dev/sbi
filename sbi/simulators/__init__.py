from sbi.simulators.linear_gaussian import diagonal_linear_gaussian, linear_gaussian
from sbi.simulators.simutils import simulate_in_batches
