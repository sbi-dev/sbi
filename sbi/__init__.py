from sbi.__version__ import __version__  # noqa: F401
