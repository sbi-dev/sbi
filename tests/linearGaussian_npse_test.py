from typing import List

import pytest
import torch
from torch import eye, ones, zeros
from torch.distributions import MultivariateNormal

from sbi import analysis as analysis
from sbi import utils as utils
from sbi.inference import NPSE
from sbi.simulators import linear_gaussian
from sbi.simulators.linear_gaussian import (
    samples_true_posterior_linear_gaussian_mvn_prior_different_dims,
    samples_true_posterior_linear_gaussian_uniform_prior,
    true_posterior_linear_gaussian_mvn_prior,
)

from .test_utils import check_c2st, get_dkl_gaussian_prior


# We always test num_dim and sample_with with defaults and mark the rests as slow.
@pytest.mark.parametrize(
    "sde_type, num_dim, prior_str, sample_with",
    [
        ("vp", 1, "gaussian", ["sde", "ode"]),
        ("vp", 3, "uniform", ["sde", "ode"]),
        ("vp", 3, "gaussian", ["sde", "ode"]),
        ("ve", 3, "uniform", ["sde", "ode"]),
        ("subvp", 3, "uniform", ["sde", "ode"]),
    ],
)
def test_c2st_npse_on_linearGaussian(
    sde_type, num_dim: int, prior_str: str, sample_with: List[str]
):
    """Test whether NPSE infers well a simple example with available ground truth."""

    x_o = zeros(1, num_dim)
    num_samples = 1000
    num_simulations = 10_000

    # likelihood_mean will be likelihood_shift+theta
    likelihood_shift = -1.0 * ones(num_dim)
    likelihood_cov = 0.3 * eye(num_dim)

    if prior_str == "gaussian":
        prior_mean = zeros(num_dim)
        prior_cov = eye(num_dim)
        prior = MultivariateNormal(loc=prior_mean, covariance_matrix=prior_cov)
        gt_posterior = true_posterior_linear_gaussian_mvn_prior(
            x_o, likelihood_shift, likelihood_cov, prior_mean, prior_cov
        )
        target_samples = gt_posterior.sample((num_samples,))
    else:
        prior = utils.BoxUniform(-2.0 * ones(num_dim), 2.0 * ones(num_dim))
        target_samples = samples_true_posterior_linear_gaussian_uniform_prior(
            x_o,
            likelihood_shift,
            likelihood_cov,
            prior=prior,
            num_samples=num_samples,
        )

    inference = NPSE(prior, sde_type=sde_type, show_progress_bars=True)

    theta = prior.sample((num_simulations,))
    x = linear_gaussian(theta, likelihood_shift, likelihood_cov)

    score_estimator = inference.append_simulations(theta, x).train(
        training_batch_size=100
    )
    # amortize the training when testing sample_with.
    for method in sample_with:
        posterior = inference.build_posterior(score_estimator, sample_with=method)
        posterior.set_default_x(x_o)
        samples = posterior.sample((num_samples,))

        # Compute the c2st and assert it is near chance level of 0.5.
        check_c2st(
            samples,
            target_samples,
            alg=f"npse-{sde_type or 'vp'}-{prior_str}-{num_dim}D-{method}",
        )

    # Checks for log_prob()
    if prior_str == "gaussian":
        # For the Gaussian prior, we compute the KLd between ground truth and
        # posterior.
        dkl = get_dkl_gaussian_prior(
            posterior,
            x_o[0],
            likelihood_shift,
            likelihood_cov,
            prior_mean,
            prior_cov,
        )

        max_dkl = 0.15

        assert dkl < max_dkl, (
            f"D-KL={dkl} is more than 2 stds above the average performance."
        )


def test_c2st_npse_on_linearGaussian_different_dims():
    """Test NPE on linear Gaussian with different theta and x dimensionality."""

    theta_dim = 3
    x_dim = 2
    discard_dims = theta_dim - x_dim

    x_o = zeros(1, x_dim)
    num_samples = 1000
    num_simulations = 2000

    # likelihood_mean will be likelihood_shift+theta
    likelihood_shift = -1.0 * ones(x_dim)
    likelihood_cov = 0.3 * eye(x_dim)

    prior_mean = zeros(theta_dim)
    prior_cov = eye(theta_dim)
    prior = MultivariateNormal(loc=prior_mean, covariance_matrix=prior_cov)
    target_samples = samples_true_posterior_linear_gaussian_mvn_prior_different_dims(
        x_o,
        likelihood_shift,
        likelihood_cov,
        prior_mean,
        prior_cov,
        num_discarded_dims=discard_dims,
        num_samples=num_samples,
    )

    def simulator(theta):
        return linear_gaussian(
            theta,
            likelihood_shift,
            likelihood_cov,
            num_discarded_dims=discard_dims,
        )

    # Test whether prior can be `None`.
    inference = NPSE(prior=None)

    theta = prior.sample((num_simulations,))
    x = simulator(theta)

    # Test whether we can stop and resume.
    inference.append_simulations(theta, x).train(
        max_num_epochs=10, training_batch_size=100
    )
    inference.train(
        resume_training=True, force_first_round_loss=True, training_batch_size=100
    )
    posterior = inference.build_posterior().set_default_x(x_o)
    samples = posterior.sample((num_samples,))

    # Compute the c2st and assert it is near chance level of 0.5.
    check_c2st(samples, target_samples, alg="npse_different_dims_and_resume_training")


@pytest.mark.xfail(
    reason="iid_bridge not working.",
    raises=AssertionError,
    strict=True,
    match="Score accumulation*",
)
@pytest.mark.parametrize("num_trials", [2, 10])
def test_npse_iid_inference(num_trials):
    """Test whether NPSE infers well a simple example with available ground truth."""

    num_dim = 2
    x_o = zeros(num_trials, num_dim)
    num_samples = 1000
    num_simulations = 3000

    # likelihood_mean will be likelihood_shift+theta
    likelihood_shift = -1.0 * ones(num_dim)
    likelihood_cov = 0.3 * eye(num_dim)

    prior_mean = zeros(num_dim)
    prior_cov = eye(num_dim)
    prior = MultivariateNormal(loc=prior_mean, covariance_matrix=prior_cov)
    gt_posterior = true_posterior_linear_gaussian_mvn_prior(
        x_o, likelihood_shift, likelihood_cov, prior_mean, prior_cov
    )
    target_samples = gt_posterior.sample((num_samples,))

    inference = NPSE(prior, show_progress_bars=True)

    theta = prior.sample((num_simulations,))
    x = linear_gaussian(theta, likelihood_shift, likelihood_cov)

    score_estimator = inference.append_simulations(theta, x).train(
        training_batch_size=100,
    )
    posterior = inference.build_posterior(score_estimator)
    posterior.set_default_x(x_o)
    samples = posterior.sample((num_samples,))

    # Compute the c2st and assert it is near chance level of 0.5.
    check_c2st(
        samples, target_samples, alg=f"npse-vp-gaussian-1D-{num_trials}iid-trials"
    )


@pytest.mark.slow
def test_npse_map():
    num_dim = 2
    x_o = zeros(num_dim)
    num_simulations = 3000

    # likelihood_mean will be likelihood_shift+theta
    likelihood_shift = -1.0 * ones(num_dim)
    likelihood_cov = 0.3 * eye(num_dim)

    prior_mean = zeros(num_dim)
    prior_cov = eye(num_dim)
    prior = MultivariateNormal(loc=prior_mean, covariance_matrix=prior_cov)
    gt_posterior = true_posterior_linear_gaussian_mvn_prior(
        x_o, likelihood_shift, likelihood_cov, prior_mean, prior_cov
    )
    inference = NPSE(prior, show_progress_bars=True)

    theta = prior.sample((num_simulations,))
    x = linear_gaussian(theta, likelihood_shift, likelihood_cov)

    inference.append_simulations(theta, x).train(
        training_batch_size=100, max_num_epochs=10
    )
    posterior = inference.build_posterior().set_default_x(x_o)

    map_ = posterior.map(show_progress_bars=True)

    assert torch.allclose(map_, gt_posterior.mean, atol=0.4), "MAP is not close to GT."
