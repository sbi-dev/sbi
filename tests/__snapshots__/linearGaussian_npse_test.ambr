# serializer version: 1
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-None_prior-dim_3-subvp]
  list([
    list([
      -0.9198087453842163,
      -0.03226412460207939,
    ]),
    list([
      -1.2253586053848267,
      -0.5765306949615479,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-None_prior-dim_3-ve]
  list([
    list([
      -0.9534644484519958,
      -0.2797377109527588,
    ]),
    list([
      -1.2012022733688354,
      -0.8256365060806274,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-None_prior-dim_3-vp]
  list([
    list([
      -0.4981342852115631,
      -0.8204771280288696,
    ]),
    list([
      -1.3586816787719727,
      -0.8610840439796448,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-gaussian_prior-dim_1-vp0]
  list([
    list([
      0.04802143573760986,
    ]),
    list([
      -0.13500557839870453,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-gaussian_prior-dim_1-vp1]
  list([
    list([
      0.04802143573760986,
    ]),
    list([
      -0.13500557839870453,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-gaussian_prior-dim_2-subvp]
  list([
    list([
      -0.9198087453842163,
      -0.03226412460207939,
    ]),
    list([
      -1.2253586053848267,
      -0.5765306949615479,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-gaussian_prior-dim_2-ve]
  list([
    list([
      -0.9534644484519958,
      -0.2797377109527588,
    ]),
    list([
      -1.2012022733688354,
      -0.8256365060806274,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-gaussian_prior-dim_2-vp]
  list([
    list([
      -0.4981342852115631,
      -0.8204771280288696,
    ]),
    list([
      -1.3586816787719727,
      -0.8610840439796448,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-uniform_prior-dim_3-subvp]
  list([
    list([
      0.5026171803474426,
      -0.14688342809677124,
      0.19819878041744232,
    ]),
    list([
      0.8101956844329834,
      0.9525981545448303,
      0.03972509503364563,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-uniform_prior-dim_3-ve]
  list([
    list([
      0.6304304599761963,
      -0.6617138385772705,
      0.1326000690460205,
    ]),
    list([
      0.8797309398651123,
      0.6335045695304871,
      -0.022958669811487198,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-ode-uniform_prior-dim_3-vp]
  list([
    list([
      0.2415347695350647,
      -0.1564069241285324,
      0.1062876507639885,
    ]),
    list([
      -0.10485473275184631,
      0.3475174605846405,
      -0.07742819935083389,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-None_prior-dim_3-subvp]
  list([
    list([
      -0.26416945457458496,
      -0.5380855202674866,
    ]),
    list([
      -0.9131697416305542,
      -0.3601299524307251,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-None_prior-dim_3-ve]
  list([
    list([
      -0.6600239276885986,
      -0.8042313456535339,
    ]),
    list([
      -0.5148065686225891,
      -0.7849245667457581,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-None_prior-dim_3-vp]
  list([
    list([
      -0.7121580243110657,
      -0.9239912629127502,
    ]),
    list([
      -1.3291730880737305,
      -0.42849549651145935,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-gaussian_prior-dim_1-vp0]
  list([
    list([
      0.0026113521307706833,
    ]),
    list([
      -0.1795785129070282,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-gaussian_prior-dim_1-vp1]
  list([
    list([
      0.0026113521307706833,
    ]),
    list([
      -0.1795785129070282,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-gaussian_prior-dim_2-subvp]
  list([
    list([
      -0.26416945457458496,
      -0.5380855202674866,
    ]),
    list([
      -0.9131697416305542,
      -0.3601299524307251,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-gaussian_prior-dim_2-ve]
  list([
    list([
      -0.6600239276885986,
      -0.8042313456535339,
    ]),
    list([
      -0.5148065686225891,
      -0.7849245667457581,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-gaussian_prior-dim_2-vp]
  list([
    list([
      -0.7121580243110657,
      -0.9239912629127502,
    ]),
    list([
      -1.3291730880737305,
      -0.42849549651145935,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-uniform_prior-dim_3-subvp]
  list([
    list([
      -0.1743609607219696,
      -0.21850894391536713,
      0.07176153361797333,
    ]),
    list([
      0.296996533870697,
      -0.25024330615997314,
      -0.12501412630081177,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-uniform_prior-dim_3-ve]
  list([
    list([
      -0.17165498435497284,
      -1.066346526145935,
      -0.19592738151550293,
    ]),
    list([
      0.17051002383232117,
      -1.7759748697280884,
      -0.12025851756334305,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[auto_gauss-sde-uniform_prior-dim_3-vp]
  list([
    list([
      0.0633796900510788,
      -1.141589641571045,
      -0.11582185328006744,
    ]),
    list([
      0.42771634459495544,
      -0.5370681881904602,
      0.23705551028251648,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-None_prior-dim_3-subvp]
  list([
    list([
      -0.9198087453842163,
      -0.03226412460207939,
    ]),
    list([
      -1.2253586053848267,
      -0.5765306949615479,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-None_prior-dim_3-ve]
  list([
    list([
      -0.9534644484519958,
      -0.2797377109527588,
    ]),
    list([
      -1.2012022733688354,
      -0.8256365060806274,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-None_prior-dim_3-vp]
  list([
    list([
      -0.4981342852115631,
      -0.8204771280288696,
    ]),
    list([
      -1.3586816787719727,
      -0.8610840439796448,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-gaussian_prior-dim_1-vp0]
  list([
    list([
      0.04802143573760986,
    ]),
    list([
      -0.13500557839870453,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-gaussian_prior-dim_1-vp1]
  list([
    list([
      0.04802143573760986,
    ]),
    list([
      -0.13500557839870453,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-gaussian_prior-dim_2-subvp]
  list([
    list([
      -0.9198087453842163,
      -0.03226412460207939,
    ]),
    list([
      -1.2253586053848267,
      -0.5765306949615479,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-gaussian_prior-dim_2-ve]
  list([
    list([
      -0.9534644484519958,
      -0.2797377109527588,
    ]),
    list([
      -1.2012022733688354,
      -0.8256365060806274,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-gaussian_prior-dim_2-vp]
  list([
    list([
      -0.4981342852115631,
      -0.8204771280288696,
    ]),
    list([
      -1.3586816787719727,
      -0.8610840439796448,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-uniform_prior-dim_3-subvp]
  list([
    list([
      0.5026171803474426,
      -0.14688342809677124,
      0.19819878041744232,
    ]),
    list([
      0.8101956844329834,
      0.9525981545448303,
      0.03972509503364563,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-uniform_prior-dim_3-ve]
  list([
    list([
      0.6304304599761963,
      -0.6617138385772705,
      0.1326000690460205,
    ]),
    list([
      0.8797309398651123,
      0.6335045695304871,
      -0.022958669811487198,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-ode-uniform_prior-dim_3-vp]
  list([
    list([
      0.2415347695350647,
      -0.1564069241285324,
      0.1062876507639885,
    ]),
    list([
      -0.10485473275184631,
      0.3475174605846405,
      -0.07742819935083389,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-None_prior-dim_3-subvp]
  list([
    list([
      -0.26416945457458496,
      -0.5380855202674866,
    ]),
    list([
      -0.9131697416305542,
      -0.3601299524307251,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-None_prior-dim_3-ve]
  list([
    list([
      -0.6600239276885986,
      -0.8042313456535339,
    ]),
    list([
      -0.5148065686225891,
      -0.7849245667457581,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-None_prior-dim_3-vp]
  list([
    list([
      -0.7121580243110657,
      -0.9239912629127502,
    ]),
    list([
      -1.3291730880737305,
      -0.42849549651145935,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-gaussian_prior-dim_1-vp0]
  list([
    list([
      0.0026113521307706833,
    ]),
    list([
      -0.1795785129070282,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-gaussian_prior-dim_1-vp1]
  list([
    list([
      0.0026113521307706833,
    ]),
    list([
      -0.1795785129070282,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-gaussian_prior-dim_2-subvp]
  list([
    list([
      -0.26416945457458496,
      -0.5380855202674866,
    ]),
    list([
      -0.9131697416305542,
      -0.3601299524307251,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-gaussian_prior-dim_2-ve]
  list([
    list([
      -0.6600239276885986,
      -0.8042313456535339,
    ]),
    list([
      -0.5148065686225891,
      -0.7849245667457581,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-gaussian_prior-dim_2-vp]
  list([
    list([
      -0.7121580243110657,
      -0.9239912629127502,
    ]),
    list([
      -1.3291730880737305,
      -0.42849549651145935,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-uniform_prior-dim_3-subvp]
  list([
    list([
      -0.1743609607219696,
      -0.21850894391536713,
      0.07176153361797333,
    ]),
    list([
      0.296996533870697,
      -0.25024330615997314,
      -0.12501412630081177,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-uniform_prior-dim_3-ve]
  list([
    list([
      -0.17165498435497284,
      -1.066346526145935,
      -0.19592738151550293,
    ]),
    list([
      0.17051002383232117,
      -1.7759748697280884,
      -0.12025851756334305,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[fnpe-sde-uniform_prior-dim_3-vp]
  list([
    list([
      0.0633796900510788,
      -1.141589641571045,
      -0.11582185328006744,
    ]),
    list([
      0.42771634459495544,
      -0.5370681881904602,
      0.23705551028251648,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-None_prior-dim_3-subvp]
  list([
    list([
      -0.9198087453842163,
      -0.03226412460207939,
    ]),
    list([
      -1.2253586053848267,
      -0.5765306949615479,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-None_prior-dim_3-ve]
  list([
    list([
      -0.9534644484519958,
      -0.2797377109527588,
    ]),
    list([
      -1.2012022733688354,
      -0.8256365060806274,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-None_prior-dim_3-vp]
  list([
    list([
      -0.4981342852115631,
      -0.8204771280288696,
    ]),
    list([
      -1.3586816787719727,
      -0.8610840439796448,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-gaussian_prior-dim_1-vp0]
  list([
    list([
      0.04802143573760986,
    ]),
    list([
      -0.13500557839870453,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-gaussian_prior-dim_1-vp1]
  list([
    list([
      0.04802143573760986,
    ]),
    list([
      -0.13500557839870453,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-gaussian_prior-dim_2-subvp]
  list([
    list([
      -0.9198087453842163,
      -0.03226412460207939,
    ]),
    list([
      -1.2253586053848267,
      -0.5765306949615479,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-gaussian_prior-dim_2-ve]
  list([
    list([
      -0.9534644484519958,
      -0.2797377109527588,
    ]),
    list([
      -1.2012022733688354,
      -0.8256365060806274,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-gaussian_prior-dim_2-vp]
  list([
    list([
      -0.4981342852115631,
      -0.8204771280288696,
    ]),
    list([
      -1.3586816787719727,
      -0.8610840439796448,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-uniform_prior-dim_3-subvp]
  list([
    list([
      0.5026171803474426,
      -0.14688342809677124,
      0.19819878041744232,
    ]),
    list([
      0.8101956844329834,
      0.9525981545448303,
      0.03972509503364563,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-uniform_prior-dim_3-ve]
  list([
    list([
      0.6304304599761963,
      -0.6617138385772705,
      0.1326000690460205,
    ]),
    list([
      0.8797309398651123,
      0.6335045695304871,
      -0.022958669811487198,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-ode-uniform_prior-dim_3-vp]
  list([
    list([
      0.2415347695350647,
      -0.1564069241285324,
      0.1062876507639885,
    ]),
    list([
      -0.10485473275184631,
      0.3475174605846405,
      -0.07742819935083389,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-None_prior-dim_3-subvp]
  list([
    list([
      -0.26416945457458496,
      -0.5380855202674866,
    ]),
    list([
      -0.9131697416305542,
      -0.3601299524307251,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-None_prior-dim_3-ve]
  list([
    list([
      -0.6600239276885986,
      -0.8042313456535339,
    ]),
    list([
      -0.5148065686225891,
      -0.7849245667457581,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-None_prior-dim_3-vp]
  list([
    list([
      -0.7121580243110657,
      -0.9239912629127502,
    ]),
    list([
      -1.3291730880737305,
      -0.42849549651145935,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-gaussian_prior-dim_1-vp0]
  list([
    list([
      0.0026113521307706833,
    ]),
    list([
      -0.1795785129070282,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-gaussian_prior-dim_1-vp1]
  list([
    list([
      0.0026113521307706833,
    ]),
    list([
      -0.1795785129070282,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-gaussian_prior-dim_2-subvp]
  list([
    list([
      -0.26416945457458496,
      -0.5380855202674866,
    ]),
    list([
      -0.9131697416305542,
      -0.3601299524307251,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-gaussian_prior-dim_2-ve]
  list([
    list([
      -0.6600239276885986,
      -0.8042313456535339,
    ]),
    list([
      -0.5148065686225891,
      -0.7849245667457581,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-gaussian_prior-dim_2-vp]
  list([
    list([
      -0.7121580243110657,
      -0.9239912629127502,
    ]),
    list([
      -1.3291730880737305,
      -0.42849549651145935,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-uniform_prior-dim_3-subvp]
  list([
    list([
      -0.1743609607219696,
      -0.21850894391536713,
      0.07176153361797333,
    ]),
    list([
      0.296996533870697,
      -0.25024330615997314,
      -0.12501412630081177,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-uniform_prior-dim_3-ve]
  list([
    list([
      -0.17165498435497284,
      -1.066346526145935,
      -0.19592738151550293,
    ]),
    list([
      0.17051002383232117,
      -1.7759748697280884,
      -0.12025851756334305,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[gauss-sde-uniform_prior-dim_3-vp]
  list([
    list([
      0.0633796900510788,
      -1.141589641571045,
      -0.11582185328006744,
    ]),
    list([
      0.42771634459495544,
      -0.5370681881904602,
      0.23705551028251648,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-None_prior-dim_3-subvp]
  list([
    list([
      -0.9198087453842163,
      -0.03226412460207939,
    ]),
    list([
      -1.2253586053848267,
      -0.5765306949615479,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-None_prior-dim_3-ve]
  list([
    list([
      -0.9534644484519958,
      -0.2797377109527588,
    ]),
    list([
      -1.2012022733688354,
      -0.8256365060806274,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-None_prior-dim_3-vp]
  list([
    list([
      -0.4981342852115631,
      -0.8204771280288696,
    ]),
    list([
      -1.3586816787719727,
      -0.8610840439796448,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-gaussian_prior-dim_1-vp0]
  list([
    list([
      0.04802143573760986,
    ]),
    list([
      -0.13500557839870453,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-gaussian_prior-dim_1-vp1]
  list([
    list([
      0.04802143573760986,
    ]),
    list([
      -0.13500557839870453,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-gaussian_prior-dim_2-subvp]
  list([
    list([
      -0.9198087453842163,
      -0.03226412460207939,
    ]),
    list([
      -1.2253586053848267,
      -0.5765306949615479,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-gaussian_prior-dim_2-ve]
  list([
    list([
      -0.9534644484519958,
      -0.2797377109527588,
    ]),
    list([
      -1.2012022733688354,
      -0.8256365060806274,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-gaussian_prior-dim_2-vp]
  list([
    list([
      -0.4981342852115631,
      -0.8204771280288696,
    ]),
    list([
      -1.3586816787719727,
      -0.8610840439796448,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-uniform_prior-dim_3-subvp]
  list([
    list([
      0.5026171803474426,
      -0.14688342809677124,
      0.19819878041744232,
    ]),
    list([
      0.8101956844329834,
      0.9525981545448303,
      0.03972509503364563,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-uniform_prior-dim_3-ve]
  list([
    list([
      0.6304304599761963,
      -0.6617138385772705,
      0.1326000690460205,
    ]),
    list([
      0.8797309398651123,
      0.6335045695304871,
      -0.022958669811487198,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-ode-uniform_prior-dim_3-vp]
  list([
    list([
      0.2415347695350647,
      -0.1564069241285324,
      0.1062876507639885,
    ]),
    list([
      -0.10485473275184631,
      0.3475174605846405,
      -0.07742819935083389,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-None_prior-dim_3-subvp]
  list([
    list([
      -0.26416945457458496,
      -0.5380855202674866,
    ]),
    list([
      -0.9131697416305542,
      -0.3601299524307251,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-None_prior-dim_3-ve]
  list([
    list([
      -0.6600239276885986,
      -0.8042313456535339,
    ]),
    list([
      -0.5148065686225891,
      -0.7849245667457581,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-None_prior-dim_3-vp]
  list([
    list([
      -0.7121580243110657,
      -0.9239912629127502,
    ]),
    list([
      -1.3291730880737305,
      -0.42849549651145935,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-gaussian_prior-dim_1-vp0]
  list([
    list([
      0.0026113521307706833,
    ]),
    list([
      -0.1795785129070282,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-gaussian_prior-dim_1-vp1]
  list([
    list([
      0.0026113521307706833,
    ]),
    list([
      -0.1795785129070282,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-gaussian_prior-dim_2-subvp]
  list([
    list([
      -0.26416945457458496,
      -0.5380855202674866,
    ]),
    list([
      -0.9131697416305542,
      -0.3601299524307251,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-gaussian_prior-dim_2-ve]
  list([
    list([
      -0.6600239276885986,
      -0.8042313456535339,
    ]),
    list([
      -0.5148065686225891,
      -0.7849245667457581,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-gaussian_prior-dim_2-vp]
  list([
    list([
      -0.7121580243110657,
      -0.9239912629127502,
    ]),
    list([
      -1.3291730880737305,
      -0.42849549651145935,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-uniform_prior-dim_3-subvp]
  list([
    list([
      -0.1743609607219696,
      -0.21850894391536713,
      0.07176153361797333,
    ]),
    list([
      0.296996533870697,
      -0.25024330615997314,
      -0.12501412630081177,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-uniform_prior-dim_3-ve]
  list([
    list([
      -0.17165498435497284,
      -1.066346526145935,
      -0.19592738151550293,
    ]),
    list([
      0.17051002383232117,
      -1.7759748697280884,
      -0.12025851756334305,
    ]),
  ])
# ---
# name: test_npse_iid_inference_snapshot[jac_gauss-sde-uniform_prior-dim_3-vp]
  list([
    list([
      0.0633796900510788,
      -1.141589641571045,
      -0.11582185328006744,
    ]),
    list([
      0.42771634459495544,
      -0.5370681881904602,
      0.23705551028251648,
    ]),
  ])
# ---
